plasmapy/diagnostics/proton_radiography.py

```diff
@@ -0,0 +1,1022 @@
+                "deflected away from the detector plane. The fields "
+                "provided may be too high to successfully radiograph "
+                "with this particle energy.",
+                RuntimeWarning,
+            )
+
+        self.x += self.v * np.outer(t, np.ones(3))
+
+        # Check that all points are now in the detector plane
+        # (Eq. of a plane is nhat*x + d = 0)
+        plane_eq = np.dot(self.x, self.det_n) + np.linalg.norm(self.detector)
+        assert np.allclose(plane_eq, np.zeros(self.nparticles_grid), atol=1e-6)
+
+    def _push(self):
+        r"""
+        Advance particles using an implementation of the time-centered
+        Boris algorithm
+        """
+        # Get a list of positions (input for interpolator)
+        pos = self.x[self.grid_ind, :] * u.m
+
+        # Update the list of particles on and off the grid
+        self.on_grid = self.grid.on_grid(pos)
+        # entered_grid is zero at the end if a particle has never
+        # entered the grid
+        self.entered_grid += self.on_grid
+
+        # Estimate the E and B fields for each particle
+        # Note that this interpolation step is BY FAR the slowest part of the push
+        # loop. Any speed improvements will have to come from here.
+        if self.field_weighting == "volume averaged":
+            Ex, Ey, Ez, Bx, By, Bz = self.grid.volume_averaged_interpolator(
+                pos, "E_x", "E_y", "E_z", "B_x", "B_y", "B_z", persistent=True,
+            )
+        elif self.field_weighting == "nearest neighbor":
+            Ex, Ey, Ez, Bx, By, Bz = self.grid.nearest_neighbor_interpolator(
+                pos, "E_x", "E_y", "E_z", "B_x", "B_y", "B_z", persistent=True,
+            )
+
+        # Create arrays of E and B as required by push algorithm
+        E = np.array(
+            [Ex.to(u.V / u.m).value, Ey.to(u.V / u.m).value, Ez.to(u.V / u.m).value]
+        )
+        E = np.moveaxis(E, 0, -1)
+        B = np.array([Bx.to(u.T).value, By.to(u.T).value, Bz.to(u.T).value])
+        B = np.moveaxis(B, 0, -1)
+
+        # Calculate the adaptive timestep from the fields currently experienced
+        # by the particles
+        # If user sets dt explicitly, that's handled in _adpative_dt
+        dt = self._adaptive_dt(Ex, Ey, Ez, Bx, By, Bz)
+
+        # TODO: Test v/c and implement relativistic Boris push when required
+        # vc = np.max(v)/_c
+
+        x = self.x[self.grid_ind, :]
+        v = self.v[self.grid_ind, :]
+        boris_push(x, v, B, E, self.q, self.m, dt)
+        self.x[self.grid_ind, :] = x
+        self.v[self.grid_ind, :] = v
+
+    def _stop_condition(self):
+        r"""
+        The stop condition is that most of the particles have entered the grid
+        and almost all have now left it.
+        """
+        # Count the number of particles who have entered, which is the
+        # number of non-zero entries in entered_grid
+        self.num_entered = np.nonzero(self.entered_grid)[0].size
```

Comment:
If `self.entered_grid` is an array of booleans, you can just do `.sum()`, no? Oh right, no, this is the counter for how many times each one has entered the grid. This looks fine, then.

Reply:
Yes, your understanding there is right.